sqlglot/optimizer/annotate_types.py

```diff
@@ -97,22 +103,47 @@ class TypeAnnotator:
         },
     }
 
+    TRAVERSABLES = (exp.Select, exp.Union, exp.UDTF, exp.Subquery)
+
     def __init__(self, schema=None, annotators=None, coerces_to=None):
         self.schema = schema
         self.annotators = annotators or self.ANNOTATORS
         self.coerces_to = coerces_to or self.COERCES_TO
 
     def annotate(self, expression):
+        scopes = traverse_scope(expression) if isinstance(expression, self.TRAVERSABLES) else []
```

Comment:
why bother with this? why not just do

if isinstance(expression, self.Traversables):
   for scope in traverse_scope(...)